File: c_files/raylib/endless_runner.c
#include "raylib.h"
#include "utils.h"

#define NUM_OF_ENEMIES 10

// PLAYER

typedef struct Player {
  Rectangle rect;
  Color color;
  float change_y;
  float gravity;
  float jump_strength;

} Player;

void player_init(Player *player) {
  player->rect = (Rectangle){100, GetScreenHeight() - 50, 20, 50};
  player->change_y = 0.0f;
  player->gravity = 1500.0f;
  player->jump_strength = -500;
  player->color = DARKGREEN;
}

void player_draw(Player *player) {
  DrawRectangleRec(player->rect, player->color);
}

void player_update(Player *player) {
  if (IsKeyPressed(KEY_UP)) {
    player->change_y = player->jump_strength;
  }

  player->change_y += player->gravity * GetFrameTime();
  player->rect.y += player->change_y * GetFrameTime();

  player->rect.y = max_float(
      0.0f, min_float(player->rect.y, GetScreenHeight() - player->rect.height));
}

// ENEMIES

typedef struct Enemy {
  Rectangle rect;
  Color color;
  float speed;
  float horizontal_distance;

} Enemy;

typedef struct Enemies {
  Enemy list[NUM_OF_ENEMIES];
  int size;

} Enemies;

void enemies_init(Enemies *enemies) { enemies->size = 0; }

void enemies_draw(Enemies *enemies) {
  if (enemies->size > 0) {
    for (int i = 0; i < enemies->size; i++) {
      DrawRectangleRec(enemies->list[i].rect, enemies->list[i].color);
    }
  }
}

void enemies_update(Enemies *enemies) {
  // add enemy
  if (enemies->size <= 0 ||
      GetScreenWidth() - enemies->list[enemies->size - 1].horizontal_distance >
          enemies->list[enemies->size - 1].rect.x +
              enemies->list[enemies->size - 1].rect.width) {

    float random_height = (float)GetRandomValue(20, 80);
    Enemy enemy = {
        .rect = (GetScreenWidth(), GetScreenHeight() - random_height,
                 GetRandomValue(10, 30), random_height),
        .color = BLACK,
        .horizontal_distance = GetRandomValue(100, 200),
        .speed = 300.0f,

    };

    enemies->list[enemies->size] = enemy;
    enemies->size += 1;
  }

  // move and remove enemy
  if (enemies->size > 0) {
    for (int i = 0; i < enemies->size; i++) {
      enemies->list[i].rect.x -= enemies->list[i].speed * GetFrameTime();

      // remove
      if (enemies->list[i].rect.x + enemies->list[i].rect.width +
              enemies->list[i].horizontal_distance <
          0) {
        enemies->size -= 1;
      }
    }
  }
}

// GAME

typedef struct Game {
  int score;
  bool game_over;

  Player player;
  Enemies enemies;

} Game;

void game_init(Game *game) {
  game->score = 0;
  game->game_over = false;

  player_init(&game->player);
  enemies_init(&game->enemies);
}

void game_draw(Game *game) {
  // draw score
  DrawText(TextFormat("Score: %d", game->score), 10, 10, 30, BLACK);

  // draw rest
  player_draw(&game->player);
  enemies_draw(&game->enemies);
}

void game_update(Game *game) {
  if (!game->game_over) {
    player_update(&game->player);
    enemies_update(&game->enemies);
  }
}

// MAIN

int main(void) {
  const int SCREEN_WIDTH = 1000;
  const int SCREEN_HEIGHT = 200;
  const char SCREEN_TITLE[] = "Endless Runner";
  const Color SCREEN_BACKGROUND = RAYWHITE;

  InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

  Game game;

  game_init(&game);

  while (!WindowShouldClose()) {
    BeginDrawing();
    ClearBackground(SCREEN_BACKGROUND);

    game_draw(&game);
    game_update(&game);

    EndDrawing();
  }

  CloseWindow();

  return 0;
}

File: python/raylib/endless_runner.py
import pyray as p
from dataclasses import dataclass

SCREEN_WIDTH = 800
SCREEN_HEIGHT = 200
SCREEN_TITLE = "Endless Runner"
SCREEN_BACKGROUND = p.SKYBLUE


class Player:
    def __init__(self):
        self.width = 20
        self.height = 80
        self.x = 100
        self.y = p.get_screen_height() - self.height
        self.color = p.RED
        self.change_y = 0
        self.gravity = 1500
        self.jump_strength = -500
        self.active = False

    def draw(self):
        p.draw_rectangle_rec(
            p.Rectangle(self.x, self.y, self.width, self.height), self.color
        )

    def update(self):
        if p.is_key_pressed(p.KeyboardKey.KEY_UP):
            self.active = True
            self.change_y = self.jump_strength

        if self.active:
            self.change_y += self.gravity * p.get_frame_time()
            self.y += self.change_y * p.get_frame_time()

        self.y = max(0, min(self.y, p.get_screen_height() - self.height))


@dataclass
class Pillar:
    x: float
    y: float
    width: float
    height: float
    scored: int


class Pillars:
    def __init__(self):
        self.list = []
        self.color = p.DARKGREEN
        self.speed = 200.0

    def draw(self):
        for pillar in self.list:
            p.draw_rectangle_rec(
                p.Rectangle(pillar.x, pillar.y, pillar.width, pillar.height),
                self.color,
            )

    def update(self):
        self.horizontal_gap = p.get_random_value(150, 250)
        if (
            len(self.list) <= 0
            or self.list[-1].x + self.list[-1].width
            < p.get_screen_width() - self.horizontal_gap
        ):
            random_height = p.get_random_value(20, 50)
            pillar = Pillar(
                width=p.get_random_value(10, 40),
                height=random_height,
                x=p.get_screen_width(),
                y=p.get_screen_height() - random_height,
                scored=False,
            )

            self.list.append(pillar)

        for pillar in self.list:
            pillar.x -= self.speed * p.get_frame_time()

            if pillar.x + pillar.width < 0:
                self.list.remove(pillar)


class Game:
    def __init__(self):
        self.score = 0
        self.game_over = False

        self.player = Player()
        self.pillars = Pillars()

    def draw(self):
        self.player.draw()
        self.pillars.draw()

        p.draw_text(f"Score: {self.score}", 10, 10, 30, p.BLACK)

        if self.game_over:
            p.draw_text(
                "GAME OVER",
                p.get_screen_width() // 2 - 120,
                p.get_screen_height() // 2 - 40,
                40,
                p.BLACK,
            )

    def update(self):
        if not self.game_over:
            self.player.update()
            self.pillars.update()

            # player collision pillars
            for pillar in self.pillars.list:
                if p.check_collision_recs(
                    p.Rectangle(
                        self.player.x,
                        self.player.y,
                        self.player.width,
                        self.player.height,
                    ),
                    p.Rectangle(pillar.x, pillar.y, pillar.width, pillar.height),
                ):
                    self.game_over = True

                if self.player.x > pillar.x + pillar.width:
                    if not pillar.scored:
                        pillar.scored = True
                        self.score += 1

        else:
            if p.is_key_pressed(p.KeyboardKey.KEY_ENTER):
                self.game_over = False
                self.score = 0

                self.player = Player()
                self.pillars = Pillars()


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    game = Game()

    while not p.window_should_close():
        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        game.draw()
        game.update()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()
